File: exchange-worker/src/repositories/exchange-rate.js
const { GetItemCommand } = require('@aws-sdk/client-dynamodb');
const { marshall, unmarshall } = require('@aws-sdk/util-dynamodb');

const { dynamodb } = require('../data-sources');
const logger = require('../utils/logger')('EXCHANGE_RATE_REPOSITORY');

const {
  EXCHANGE_TABLE,
} = process.env;

async function findExchangeRateById(id) {
  try {
    const result = await dynamodb.client.send(new GetItemCommand({
      TableName: EXCHANGE_TABLE,
      Key: marshall({
        pk: id,
        sk: 'EXCHANGE_RATE',
      }),
    }));

    return result.Item ? unmarshall(result.Item) : null;
  } catch (error) {
    logger.error({
      message: 'Failed to find exchange rate on dynamodb',
      error_message: error.message,
      error_stack: error.stack,
    });

    throw error;
  }
}

module.exports = {
  findExchangeRateById,
};

File: exchange-api/repositories/exchange-rate.go
package repositories

import (
	"context"
	"exchange-api/entities"
	"exchange-api/utils/logger"
	"os"

	"github.com/aws/aws-sdk-go/aws"
	"github.com/aws/aws-sdk-go/service/dynamodb"
)

type (
	ExchangeRateRepositoryInterface interface {
		CreateExchangeRate(ctx context.Context, payload *entities.ExchangeRate) (*entities.ExchangeRate, error)
	}

	ExchangeRateRepository struct {
		dynamoClient *dynamodb.DynamoDB
	}
)

var exchangeRateLogger = logger.NewLogger()
var exchangeTableName = os.Getenv("EXCHANGE_TABLE")

func NewExchangeRateRepository(dynamo *dynamodb.DynamoDB) ExchangeRateRepositoryInterface {
	return ExchangeRateRepository{
		dynamoClient: dynamo,
	}
}

func (r ExchangeRateRepository) CreateExchangeRate(ctx context.Context, exchangeRate *entities.ExchangeRate) (*entities.ExchangeRate, error) {
	requestId := ctx.Value("requestId").(string)

	input := &dynamodb.PutItemInput{
		Item: map[string]*dynamodb.AttributeValue{
			"pk": {
				S: aws.String(exchangeRate.Id),
			},
			"sk": {
				S: aws.String("EXCHANGE_RATE"),
			},
			"currency_id": {
				S: aws.String(exchangeRate.CurrencyId),
			},
			"base_currency_id": {
				S: aws.String(exchangeRate.BaseCurrencyId),
			},
			"rate": {
				S: aws.String(exchangeRate.Rate),
			},
			"timestamp": {
				S: aws.String(exchangeRate.Timestamp),
			},
		},
		TableName: aws.String(exchangeTableName),
	}

	_, err := r.dynamoClient.PutItem(input)

	if err != nil {
		exchangeRateLogger.Error(map[string]interface{}{
			"message":       "Failed to create exchange in dynamodb",
			"error_message": err.Error(),
			"request_id":    requestId,
		})

		return nil, err
	}

	return exchangeRate, nil
}
